lib/api.rb

```diff
@@ -128,13 +136,14 @@ def validation_error(msg)
         }.to_json ]
       end
 
-      d = Device.find_by_uuid(params[:deviceidentifier])
-      if d && d.user_uuid != u.uuid
-        # wat
-        d.destroy
-        d = nil
+      if params[:deviceidentifier]
+        d = Device.find_by_uuid(params[:deviceidentifier])
+        if d && d.user_uuid != u.uuid
+          # wat
+          d.destroy
+          d = nil
+        end
```

Comment:
A previous chunk removed the enforcement of `:deviceidentifier`, `:devicename`, and `:devicetype` via `need_params()` but when `:deviceidentifier` is present (from non-web-vault requests), you need to still enforce that `:devicename` and `:devicetype` are present, and move the later assignments of them to `d.type` and `d.name` into this block after enforcement.

Reply:
@Pschittt I think it's expected that you reply to this, somehow ;)

Reply:
@andrzejwp
I’ve updated the code to Check that devicename and devicetype are present.
:)

Reply:
I'm still not seeing where you're enforcing that these parameters are there.

Reply:
I'm still not seeing where you're enforcing that these parameters are there.

Reply:
if statement on line 176, line 190, line 194

Reply:
That doesn't enforce that the parameters are there, that just conditionally uses them if they are.  When they were in  `need_params`, the processing would stop and return an error if the parameters weren't there.

Reply:
But if you enforce these 2 parameters, that breaks the web vault compatibility, as the web vault doesn't send them.

We can only check if they are here to use them.

Reply:
One can distinguish between web-vault requests and app-requests, and then require those two parameters or not. This should be possible with something like this:

```
    def web_vault_request?
      !!(params[:client_id].to_s == "web")
    end
```